File: V1/backend/app/routers/transactions.py
from typing import List

from app.core.controller.Transactions import TransactionsController
from app.core.schema.Responses.Error import ErrorModel, Errors
from app.core.schema.Responses.Success import Success
from app.core.schema.Transactions.Request import (
    CreateRequest,
    DeleteRequest,
    UpdateRequest,
)
from app.core.schema.Transactions.Response import Response
from fastapi import APIRouter, status
from fastapi.responses import ORJSONResponse
from pymongo.errors import DuplicateKeyError

transactionsRouter = APIRouter(prefix="/transactions")
transactionsController = TransactionsController()


@transactionsRouter.get(
    "/",
    response_model=List[Response],
    status_code=status.HTTP_200_OK,
    tags=["Transactions-Controller"],
    dependencies=None,
    summary="Get All transactions",
    description="Get all transactions for user",
    response_description="Successful Response",
    responses={
        Success.HTTP_200_OK.status_code: {
            "model": List[Response],
            "description": "Transactions found",
        },
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
    },
)
async def read_transaction_list(user_id: str) -> List[Response] | ORJSONResponse:
    if transaction := await transactionsController.get_all(user_id=user_id):
        return transaction
    else:
        return Errors.HTTP_404_NOT_FOUND


@transactionsRouter.post(
    "/",
    response_model=Response,
    status_code=status.HTTP_201_CREATED,
    tags=["Transactions-Controller"],
    dependencies=None,
    summary="Create Transaction",
    description="Create a new transaction in db",
    response_description="Successful Response",
    responses={
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
        Errors.HTTP_409_ALREADY_EXISTS.status_code: {
            "model": ErrorModel,
            "description": "Transaction already exists",
        },
        status.HTTP_201_CREATED: {
            "model": Response,
            "description": "Successfully Created",
        },
    },
)
async def create_transaction(transaction: CreateRequest) -> Response | ORJSONResponse:
    try:
        return await transactionsController.create_transaction(
            user_id=transaction.user_id,
            transaction_date=transaction.transactionDate,
            transaction_type=transaction.transactionType,
            sub_transaction_type=transaction.subTransactionType,
            debit_account=transaction.debitAccount,
            credit_account=transaction.creditAccount,
            category=transaction.category,
            description=transaction.description,
            amount=transaction.amount,
        )
    except DuplicateKeyError:
        return Errors.HTTP_409_ALREADY_EXISTS


@transactionsRouter.put(
    "/",
    response_model=Response,
    status_code=status.HTTP_200_OK,
    tags=["Transactions-Controller"],
    dependencies=None,
    summary="Update Transaction",
    description="Update the transaction based on id",
    response_description="Successful Response",
    responses={
        Success.HTTP_200_OK.status_code: {
            "model": Response,
            "description": "Transaction updated",
        },
        Errors.HTTP_404_NOT_FOUND.status_code: {
            "model": ErrorModel,
            "description": "Transaction not found",
        },
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
    },
)
async def update_transaction(transaction: UpdateRequest) -> Response | ORJSONResponse:
    if update_transaction := await transactionsController.update_transaction(
        transactionId=transaction.transactionId,
        user_id=transaction.user_id,
        transaction_date=transaction.transactionDate,
        transaction_type=transaction.transactionType,
        sub_transaction_type=transaction.subTransactionType,
        debit_account=transaction.debitAccount,
        credit_account=transaction.creditAccount,
        category=transaction.category,
        description=transaction.description,
        amount=transaction.amount,
    ):
        return update_transaction
    else:
        return Errors.HTTP_404_NOT_FOUND


@transactionsRouter.delete(
    "/",
    response_model=None,
    status_code=status.HTTP_204_NO_CONTENT,
    tags=["Transactions-Controller"],
    dependencies=None,
    summary="Delete Transaction",
    description="Delete the transaction based on id",
    response_description="Successful Response",
    responses={
        Success.HTTP_204_NO_CONTENT: {
            "model": None,
            "description": "Transaction deleted successfully",
        },
        Errors.HTTP_404_NOT_FOUND.status_code: {
            "model": ErrorModel,
            "description": "Transaction not found",
        },
        Errors.HTTP_500_INTERNAL_SERVER_ERROR.status_code: {
            "model": ErrorModel,
            "description": "Generic Error Occured",
        },
    },
)
async def delete_todo(transaction: DeleteRequest) -> ORJSONResponse | int:
    deleted_account = await transactionsController.delete_transaction(
        transactionId=transaction.transactionId, user_id=transaction.user_id
    )

    if deleted_account is False:
        return Errors.HTTP_404_NOT_FOUND

    return Success.HTTP_204_NO_CONTENT

File: frontend/src/services/transactions.ts
import axios from 'axios';
import { API_URL } from '../constants/api';
import { ErrorResponseType } from '../types/Response';
import {
  TransactionAddType,
  TransactionDeleteType,
  TransactionsType,
  TransactionsUpdateType
} from '../types/Transactions';

export async function getTransactions(
  user_id: string
): Promise<{ transactions: TransactionsType[] } | ErrorResponseType> {
  try {
    const { data, status } = await axios.get(API_URL + '/transactions?user_id=' + user_id);

    if (status === 200) {
      console.info('Transactions lists fetched');
      return { transactions: data };
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error.message);
      return { error: error.message, statusCode: error.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}

export async function addTransaction(new_account: TransactionAddType): Promise<TransactionsType | ErrorResponseType> {
  const body = {
    user_id: new_account.user_id,
    transactionDate: new_account.transactionDate,
    transactionType: new_account.transactionType,
    subTransactionType: new_account.subTransactionType,
    debitAccount: new_account.debitAccount,
    creditAccount: new_account.creditAccount,
    category: new_account.category,
    description: new_account.description,
    amount: new_account.amount
  };

  try {
    const { data, status } = await axios.post(API_URL + '/transactions', body);

    if (status === 201) {
      console.info('Transaction added successfully');
      return data;
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error);
      return { error: error.response.data, statusCode: error.response.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}

export async function updateTransaction(
  updateTransaction: TransactionsUpdateType
): Promise<TransactionsType | ErrorResponseType> {
  const body = {
    transactionId: updateTransaction.transactionId,
    user_id: updateTransaction.user_id,
    transactionDate: updateTransaction.transactionDate,
    transactionType: updateTransaction.transactionType,
    subTransactionType: updateTransaction.subTransactionType,
    debitAccount: updateTransaction.debitAccount,
    creditAccount: updateTransaction.creditAccount,
    category: updateTransaction.category,
    description: updateTransaction.description,
    amount: updateTransaction.amount
  };

  try {
    const { data, status } = await axios.put(API_URL + '/transactions', body);

    if (status === 200) {
      console.info('Transaction updated successfully');
      return data;
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error);
      return { error: error.response.data, statusCode: error.response.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}

export async function deleteTransaction(
  deleteTransaction: TransactionDeleteType
): Promise<boolean | ErrorResponseType> {
  const body = {
    transactionId: deleteTransaction.transactionId,
    user_id: deleteTransaction.user_id
  };
  try {
    const { status } = await axios.delete(API_URL + '/transactions', { data: body });

    if (status === 204) {
      console.info('Transaction deleted');
      return true;
    } else {
      console.error('Unexpected error status code:', status);
      return { error: 'An unexpected error occurred', statusCode: status };
    }
  } catch (error) {
    if (axios.isAxiosError(error)) {
      console.error('error message: ', error.message);
      return { error: error.message, statusCode: error.status };
    } else {
      console.error('unexpected error: ', error);
      return { error: 'An unexpected error occurred' };
    }
  }
}
